lib/graphql/batch/executor.rb

```diff
@@ -0,0 +1,38 @@
+module GraphQL::Batch
+  class Executor
+    THREAD_KEY = :"#{name}.batched_queries"
+    private_constant :THREAD_KEY
+
+    def self.current
+      Thread.current[THREAD_KEY] ||= new
+    end
+
+    attr_reader :loaders
+
+    def initialize
+      @loaders = {}
+    end
+
+    def shift
+      @loaders.shift.last
+    end
+
+    def execute(loader)
```

Comment:
I started to split up the methods, since I plan on allowing a custom Executor to be specified.  For instance, a multi-threaded executor would want to add locking around `shift`, and would want to be able to replace `wait` so it can spawn multiple threads to execute multiple loaders in parallel.

I guess `execute` could probably move to the `Loader` class, but I don't want the custom Loader class to have to do this error handling, so I'll have to choose another appropriate name.